CONTRIBUTING.md

```diff
@@ -0,0 +1,142 @@
+  broadly compatible with what our existing prevailing style was. We don't
+  require that everyone has clang-format installed - the weekly cleaning job
+  will pick it up if you don't - but if you do, remember to run with
+  `-style=file` to pick up our configuration.
+
+### Less strict guidelines
+
+These should be followed wherever possible - but remember the [Zen of Python]'s
+"Practicality beats purity.", and [PEP8]'s "A Foolish Consistency is the
+Hobgoblin of Little Minds" - it's okay to stray from the rules if you have a
+good reason for it, but "I prefer it this way" isn't a strong enough reason -
+there is real value to a standard style, because diverging from it sends a
+message that the code in question is special and care should be taken.
+
+- **Try to write code compatible with python 2 and 3** - we have both
+  [six] and [future] available for help with this, but try to use future to
+  write idiomatic python 3 wherever possible. If you are unsure how to do this,
+  ask for help. Once we are fully python 3 compatible, this will become a
+  requirement, but for now we don't expect most people to be developing on top
+  of python 3. If you get used to it now, it'll be less of a shock when we
+  require and test it.
+- Err on the side of [PEP8] when making any style decision not covered by
+  black. In particular, **use PEP8 as a guide for naming** when you aren't sure
+  the correct form to use.
+- **All imports should go at the top of the file** - not inline in a function.
+  This makes it very easy to reason about the dependencies of a module, makes
+  it hard to make conflicting definitions, and avoids duplicate imports
+  scattered throughout the file (one file used to have over 30 separate imports
+  of the `sys` package). If an import is for an optional dependency, then it
+  should be imported in a try/except block, with some fallback to identify the
+  missing case (such as setting the name to `None`. **Matplotlib is an
+  exception** to this rule - because it's startup logic defines the backend it
+  uses, this is permitted to be imported inline. There are still a few
+  exceptions to help avoid circular imports that are hard to remove. Anything
+  that is a python standard library import should be at the top of the file,
+  without question. However, _you don't need to fix this for existing code_.
+- **Don't create classes with one function**. If your class is an `__init__`
+  and a single 'do' function (or a `__call__`), then it can probably be more
+  concisely expressed as a function.
+- Avoid classes that work by `__call__` unless you have a good reason for it to
+  act as a functor. A named function to do the action instead is almost always
+  clearer with a proper name.
+- **Write docstrings**. We have a mix of styles at the moment, but new
+  docstrings should try to follow [Google-style] - it has a good balance
+  between clarity and length.
```

Comment:
After reading 'clean code' I am now less convinced this is a good thing than I was before.

Reply:
Docstrings in general, or this style in particular? It's probably google/numpy/sphinx, and sphinx is almost unreadable for in-line code (where you read most of them). Somewhat less useful if using type annotations but there's still many complex and interestingly-parametered functions out there

Reply:
docstrings in general. ie. no docstring is better than
```python
def wibblyfy(sprocket):
  """a function to wibblyfy a sprocket

  Args:
    sprocket (object): a sprocket

  Returns: 
    object: a wibblified sprocket or something else depending on the sprocket wibblification outcome
  """
```

Reply:
I agree, it should be wrapped to column limit.....

In those cases where it truly is entirely obvious and redundant (which probably matches small do-one-thing-well unit testable functions) I'd argue that it would happily come under "Having a good reason to veer from the guidelines".

Also, even though most of the time the code will be read, I still think the goal of having readable html-generated documentation is a good idea, especially since we have been actively working/talking about tidying up the API interfaces - in which case everything exposed as an API should have this anyway (but maybe not needing the full formal structure). Many internal functions probably _could_ get away with the one-line docstring.

(Typing and type annotations is a separate discussion for when we are fully python 3, but I think much of the same argument applies)

Reply:
having said that - I do agree with the idea of settling on one format. And since we do not have type annotations [in 99% of cases. I am aware of your leadership role in this, @jbeilstenedmands :)] these comments are useful. I did notice this in my recent dxtbx work and a few more comments would likely have made that job easier. So let's keep it as it is, and reconsider this particular can of worms once we have dropped Python 2 and after we have established a culture of type annotation :)